src/Application/Items/Commands/UpgradeUserItemCommand.cs

```diff
@@ -0,0 +1,101 @@
+using AutoMapper;
+using Crpg.Application.Common.Interfaces;
+using Crpg.Application.Common.Mediator;
+using Crpg.Application.Common.Results;
+using Crpg.Application.Common.Services;
+using Crpg.Application.Items.Models;
+using Crpg.Domain.Entities.Items;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+using LoggerFactory = Crpg.Logging.LoggerFactory;
+
+namespace Crpg.Application.Items.Commands;
+
+public record UpgradeUserItemCommand : IMediatorRequest<UserItemViewModel>
+{
+    public int UserItemId { get; init; }
+    public int UserId { get; init; }
+
+    internal class Handler : IMediatorRequestHandler<UpgradeUserItemCommand, UserItemViewModel>
+    {
+        private static readonly ILogger Logger = LoggerFactory.CreateLogger<UpgradeUserItemCommand>();
+
+        private readonly ICrpgDbContext _db;
+        private readonly IMapper _mapper;
+        private readonly IActivityLogService _activityLogService;
+
+        public Handler(ICrpgDbContext db, IMapper mapper, IActivityLogService activityLogService)
+        {
+            _db = db;
+            _mapper = mapper;
+            _activityLogService = activityLogService;
+        }
+
+        public async Task<Result<UserItemViewModel>> Handle(UpgradeUserItemCommand req, CancellationToken cancellationToken)
+        {
+            var user = await _db.Users
+                .Include(u => u.Items)
+                .ThenInclude(ui => ui.Item)
+                .AsSplitQuery()
+                .FirstOrDefaultAsync(u => u.Id == req.UserId, cancellationToken);
+
+            if (user == null)
+            {
+                return new(CommonErrors.UserNotFound(req.UserId));
+            }
+
+            if (user.HeirloomPoints < 1)
+            {
+                return new(CommonErrors.NotEnoughHeirloomPoints(1, user.HeirloomPoints));
+            }
+
+            var userItemToUpgrade = user.Items
+                .FirstOrDefault(ui => ui.Id == req.UserItemId);
+
+            if (userItemToUpgrade == null)
+            {
+                return new(CommonErrors.UserItemNotFound(req.UserItemId));
+            }
+
+            if (userItemToUpgrade.IsBroken)
+            {
+                return new(CommonErrors.ItemBroken(userItemToUpgrade.Id.ToString()));
```

Comment:
The id of the item should be used here

Reply:
@verdie-g 
Useritem.id or UserItem.ItemId , or UserItem.Item.Id, or UserItem.Item.Itemid?

Reply:
Not the user item id but the item id, that is `userItemToUpgrade.ItemId`